src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Properties.jl

```diff
@@ -654,6 +650,55 @@ end
 is_smooth(X::AbsSpec{<:Field, <:MPolyRing}) = true
 is_smooth(X::AbsSpec{<:Field, <:MPolyLocRing}) = true
 
-@attr function is_integral(X::AbsSpec)
-  return Oscar._is_integral_domain(OO(X))
+###################################################################
+# Irreducibility and Integrality                                  #
+#    integral iff (irreducible && reduced)                        #
+#    integral = OO(X) is integral domain                          #
+#    irreducible = nilradical of OO(X) is prime                   #
+###################################################################
+@doc Markdown.doc"""
+   is_irreducible(X::AbsSpec)
+
+Return the boolean value whether an affine scheme `X` is irreducible.
+
+!!! note
+    Irreducibility is checked over the (computable) base field of the affine scheme as specified upon creation of the ring, not over the algebraic closure thereof.
+
+"""
+@attr function is_irreducible(X::AbsSpec{<:Field, <:MPolyAnyRing})
+  !is_empty(X) || return false
+  !get_attribute(X, :is_integral, false) || return true 
+                                           ## integral = irreducible + reduced
+  return (length(minimal_primes(saturated_ideal(modulus(OO(X))))) == 1)
+end
+
+@doc Markdown.doc"""
+   is_integral(X::AbsSpec)
+
+Return the boolean value whether an affine scheme `X` is integral, i.e. irreducible and reduced.
+
+"""
+@attr function is_integral(X::AbsSpec{<:Field, <:MPolyAnyRing})
+  !is_empty(X) || return false
+  if has_attribute(X,:is_reduced) && has_attribute(X,:is_irreducible)
+     return get_attribute(X,:is_reduced) && get_attribute(X,:is_irreducible)
+  end
+  return is_prime(modulus(OO(X)))
+end
+
+###################################################################
+# Connectedness                                                   #
+###################################################################
+@doc Markdown.doc"""
+   is_connected(X::AbsSpec)
+
+Return the boolean value whether an affine scheme `X` is connected.
+
+"""
+@attr function is_connected(X::AbsSpec)
+  has_attribute(X,:is_connected) || error("not implemented yet")
```

Comment:
But this line is *only* executed when the `is_connected` attribute is not set. So you might as well do this:
```suggestion
  error("not implemented yet")
```

Reply:
If someone (or some other function) set the attribute, we want to return it. So the line is precisely, what was intended.

Reply:
If someone else sets the attribute, this code won't be reached. That's the point pf `@attr` !

Reply:
Thanks for that pointer. This is precisely, what was intended.